python/ray/data/datasource/mongo_datasource.py

```diff
@@ -0,0 +1,185 @@
+        >>> ds = ray.data.read_datasource( # doctest: +SKIP
+        ...     MongoDatasource(),
+        ...     uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin", # noqa: E501,
+        ...     database="my_db",
+        ...     collection="my_collection",
+        ...     schema=Schema({"col1": pa.string(), "col2": pa.int64()}),
+        ... )
+    """
+
+    def create_reader(self, **kwargs) -> Reader:
+        return _MongoDatasourceReader(**kwargs)
+
+    def do_write(
+        self,
+        blocks: List[ObjectRef[Block]],
+        metadata: List[BlockMetadata],
+        ray_remote_args: Optional[Dict[str, Any]],
+        uri: str,
+        database: str,
+        collection: str,
+    ) -> List[ObjectRef[WriteResult]]:
+        _validate_database_collection_exist(
+            pymongo.MongoClient(uri), database, collection
+        )
+
+        def write_block(uri: str, database: str, collection: str, block: Block):
+            from pymongoarrow.api import write
+
+            client = pymongo.MongoClient(uri)
+            write(client[database][collection], block)
+
+        if ray_remote_args is None:
+            ray_remote_args = {}
+
+        write_block = cached_remote_fn(write_block).options(**ray_remote_args)
+        write_tasks = []
+        for block in blocks:
+            write_task = write_block.remote(uri, database, collection, block)
+            write_tasks.append(write_task)
+        return write_tasks
+
+
+class _MongoDatasourceReader(Reader):
+    def __init__(
+        self,
+        uri: str,
+        database: str,
+        collection: str,
+        pipeline: List[Dict] = None,
+        schema: Schema = None,
+        **mongo_args,
+    ):
+        self._uri = uri
+        self._database = database
+        self._collection = collection
+        self._pipeline = pipeline
+        self._schema = schema
+        self._mongo_args = mongo_args
+        # If pipeline is unspecified, read the entire collection.
+        if not pipeline:
+            self._pipeline = [{"$match": {"_id": {"$exists": "true"}}}]
+
+        self._client = pymongo.MongoClient(uri)
+        _validate_database_collection_exist(self._client, database, collection)
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        # TODO(jian): Add memory size estimation to improve auto-tune of parallelism.
+        return None
+
+    def _get_match_query(self, pipeline: List[Dict]) -> Dict:
+        if len(pipeline) == 0 or "$match" not in pipeline[0]:
+            return {}
+        return pipeline[0]["$match"]
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        from bson.objectid import ObjectId
+
+        coll = self._client[self._database][self._collection]
+        match_query = self._get_match_query(self._pipeline)
+        partitions_ids = list(
+            coll.aggregate(
+                [
+                    {"$match": match_query},
```

Comment:
if `match_query` is not empty, will this aggregate become heavy-weight that needs read the actual data in documents?

Based on https://www.mongodb.com/docs/manual/reference/operator/aggregation/bucketAuto/, I think `$bucketAuto` should be light weight to only read metadata in colleciton. I am not sure whether `$bucketAuto` combined with `$match` is still light-weight or not.

Is it possible to verify in a large collection?